rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -131,21 +132,37 @@ class QOSEventHandler : public QOSEventHandlerBase
     }
   }
 
-  /// Execute any entities of the Waitable that are ready.
+  /// Take data so that the callback cannot be called again
   void
-  execute() override
+  take_data(std::shared_ptr<void> & data) override
   {
-    EventCallbackInfoT callback_info;
+    if (data) {
+      throw std::runtime_error("Should not be data in the pointer");
+    }
 
+    EventCallbackInfoT callback_info;
     rcl_ret_t ret = rcl_take_event(&event_handle_, &callback_info);
     if (ret != RCL_RET_OK) {
       RCUTILS_LOG_ERROR_NAMED(
         "rclcpp",
         "Couldn't take event info: %s", rcl_get_error_string().str);
       return;
     }
+    data = std::static_pointer_cast<void>(
+      std::make_shared<EventCallbackInfoT>(
+        callback_info));
+  }
 
-    event_callback_(callback_info);
+  /// Execute any entities of the Waitable that are ready.
+  void
+  execute(std::shared_ptr<void> & data) override
+  {
+    if (!data) {
+      throw std::runtime_error("Data is empty");
```

Comment:
this wasn't raising an exception before (see https://github.com/ros2/rclcpp/pull/1241/commits/70a55c505254651e92bacac76f19f8acc71cd6bf), but I guess the change is ok.

Reply:
and https://github.com/ros2/rclcpp/pull/1241/commits/7ff69fad758a68f146d065dadbba9969cb4c9bb9